module/os/linux/zfs/zpl_file.c

```diff
@@ -127,17 +127,29 @@ static int
 zpl_fsync(struct file *filp, int datasync)
 {
 	struct inode *inode = filp->f_mapping->host;
+	znode_t *zp = ITOZ(inode);
+	zfsvfs_t *zfsvfs = ITOZSB(inode);
 	cred_t *cr = CRED();
 	int error;
 	fstrans_cookie_t cookie;
 
+	ZPL_ENTER(zfsvfs);
+	ZPL_VERIFY_ZP(zp);
+	atomic_inc_32(&zp->z_sync_writes_cnt);
+	ZPL_EXIT(zfsvfs);
```

Comment:
@behlendorf iirc, i had considered this but then we run into a double-counting issue in the range zpl_fsync method (when HAVE_FSYNC_RANGE is defined): 
https://github.com/shaan1337/zfs/blob/f02f991a52cb99db2ae565bfe684f359222e58df/module/os/linux/zfs/zpl_file.c#L173

In this method, it's important to increment `z_sync_writes_cnt` before entering `filemap_write_and_wait_range()`.

However, if we adopt the simple way, it seems that it will still work fine if we just decrement `z_sync_writes_cnt` after the call to `filemap_write_and_wait_range()` and before calling `zfs_fsync()` (where it'll be incremented and decremented again).

Reply:
> it seems that it will still work fine if we just decrement z_sync_writes_cnt after the call to filemap_write_and_wait_range() and before calling zfs_fsync()

Right, this seems preferable to me.

Reply:
fixed